deploy-service/common/src/main/java/com/pinterest/deployservice/handler/PingHandler.java

```diff
@@ -517,4 +519,52 @@ PingResponseBean generateDeleteResponse(GoalAnalyst.UninstallCandidate candidate
         response.setDeployGoal(goal);
         return response;
     }
+
+
+    private static boolean isApplicable(Integer val) {
+        return val != null && val > 0;
+    }
+
+    /**
+     * Get the parallel count for execution. There are two configs can have effect. The maxParallel that specifies
+     * how many hosts and maxParallelPercentage that specifies the percentage of the Hosts. In common cases, there
+     * should be only one of them be effective as the UI and Bean update clear the other (Set to 0) when you set one.
+     * The rules are simple:
+     * 1. If only one of them is set, use that.
+     * 2. If both of them are set, use the minimum value
+     * @param environBean
+     * @param totalHosts
+     * @return The maximum parallel count the system respects
+     * @throws Exception
+     */
+    public final static int getFinalMaxParallelCount(EnvironBean environBean, long totalHosts) throws Exception {
+
+        int ret = Constants.DEFAULT_MAX_PARALLEL_HOSTS;
+        LOG.info("Get final maximum parallel count for total capactiy {}", totalHosts);
+        boolean numIsApplicable = isApplicable(environBean.getMax_parallel());
+        boolean percentageIsApplicable = isApplicable(environBean.getMax_parallel_deploy_percentage());
+
+        if (!numIsApplicable && percentageIsApplicable) {
+            ret = (int) (totalHosts * environBean.getMax_parallel_deploy_percentage() / 100);
+            LOG.info("Max parallel count {} is decided by percentage solely. Percentage is {}",
+                    ret,
+                    environBean.getMax_parallel_deploy_percentage());
+        } else if (!percentageIsApplicable && numIsApplicable) {
+            ret = environBean.getMax_parallel();
+            LOG.info("Max parallel count {} is decided by number of host solely.", ret);
+        } else if (numIsApplicable && percentageIsApplicable) {
+            ret = Math.min(environBean.getMax_parallel(),
+                    (int) (totalHosts * environBean.getMax_parallel_deploy_percentage() / 100));
+            LOG.info("Max parallel count {} is decided by combing host {} and percentage {}.",
+                    ret, environBean.getMax_parallel(), environBean.getMax_parallel_deploy_percentage());
+        }
+
+        //Ensure the value falls into the range making sense
+        if (ret <= 0) {
+            ret = 1;
+        } else if (ret > totalHosts) {
+            ret = (int) totalHosts;
+        }
+        return ret;
```

Comment:
a suggestion, regardless of applicable or not, we could just compute both the totals and return the min(). I think it will make the code much simpler and easy to read.